File: integertoenglishwords.java
//Java Solution

class Solution {
    public String numberToWords(int num) {
        Map<Integer, String> map = new HashMap<>();
        map.put(1000000000, "Billion");
        map.put(1000000, "Million");
        map.put(1000, "Thousand");
        map.put(100, "Hundred");
        map.put(90, "Ninety");
        map.put(80, "Eighty");
        map.put(70, "Seventy");
        map.put(60, "Sixty");
        map.put(50, "Fifty");
        map.put(40, "Forty");
        map.put(30, "Thirty");
        map.put(20, "Twenty");
        map.put(19, "Nineteen");
        map.put(18, "Eighteen");
        map.put(17, "Seventeen");
        map.put(16, "Sixteen");
        map.put(15, "Fifteen");
        map.put(14, "Fourteen");
        map.put(13, "Thirteen");
        map.put(12, "Twelve");
        map.put(11, "Eleven");
        map.put(10, "Ten");
        map.put(9, "Nine");
        map.put(8, "Eight");
        map.put(7, "Seven");
        map.put(6, "Six");
        map.put(5, "Five");
        map.put(4, "Four");
        map.put(3, "Three");
        map.put(2, "Two");
        map.put(1, "One");
        
        List<Pair<Integer, String>> sorted = new ArrayList<>();
        for (Map.Entry<Integer, String> set : map.entrySet()) {
            sorted.add(new Pair(set.getKey(), set.getValue()));
        }
        Collections.sort(sorted, (a,b) -> b.getKey() - a.getKey());
        
        if (num == 0) {
            return "Zero";
        }
        
        if (num <= 20) {
            return map.get(num);
        }
        
        List<String> result = new ArrayList<>();
        
        for (int i = 0; i < sorted.size(); i++) {
            int value = sorted.get(i).getKey();
            String translation = sorted.get(i).getValue();
            
            int repeat = num / value;
            
            if (repeat == 0) {
                continue;
            }
            
            num -= (repeat * value);
            
            if (repeat == 1 && value >= 100) {
                result.add("One");
                result.add(translation);
                continue;
            }
            
            if (repeat == 1) {
                result.add(translation);
                continue;
            }
            
            result.add(numberToWords(repeat));
            result.add(translation);
        }
        
        String[] sentence = new String[result.size()];
        for (int i = 0; i < result.size(); i++) {
            sentence[i] = result.get(i);
        }
        
        return String.join(" ", sentence);
    }
}

File: integertoenglishwords.js
//Objective is to convert an integer into english words

let num = 12345


//O(n) solution that uses a hashmap to keep track of all value -> word pairs

let translations = new Map([
    [1000000000, 'Billion'],
    [1000000, 'Million'],
    [1000, 'Thousand'],
    [100, 'Hundred'],
    [90, 'Ninety'],
    [80, 'Eighty'],
    [70, 'Seventy'],
    [60, 'Sixty'],
    [50, 'Fifty'],
    [40, 'Forty'],
    [30, 'Thirty'],
    [20, 'Twenty'],
    [19, 'Nineteen'],
    [18, 'Eighteen'],
    [17, 'Seventeen'],
    [16, 'Sixteen'],
    [15, 'Fifteen'],
    [14, 'Fourteen'],
    [13, 'Thirteen'],
    [12, 'Twelve'],
    [11, 'Eleven'],
    [10, 'Ten'],
    [9, 'Nine'],
    [8, 'Eight'],
    [7, 'Seven'],
    [6, 'Six'],
    [5, 'Five'],
    [4, 'Four'],
    [3, 'Three'],
    [2, 'Two'],
    [1, 'One'],
])

if (num == 0) {
    return 'Zero'
}

if (num <= 20) {
    return translations.get(num)
}

let result = []

for (let [value, translation] of translations) {
    let repeats = Math.floor(num / value)
    
    //Too large, find smaller value
    if (repeats == 0) {
        continue
    }
    
    //Subtract the amount from the number
    num -= repeats * value
    
    //'One' Hundred, 'One' Thousand and so on...
    //Make sure not to call the next if statement
    if (repeats == 1 && value >= 100) {
        result.push('One', translation)
        continue
    }
    
    //The value is less than 100 and only repeats once
    if (repeats == 1) {
        result.push(translation)
        continue
    }
    
    //Make a recursive call to the 'repeats' to get the translation for it
    result.push(numberToWords(repeats), translation)
}

return result.join(' ')